### ISSUE TITLE
AWS_MSK_IAM: kept getting GROUP_AUTHORIZATION_FAILED after several hours of running

### ISSUE THREAD

**Issue by gunturaf** [twmb/franz-go#147]
I use AWS_MSK_IAM for the SASL method, I also set the AWS credentials to also be cached using [NewCredentialsCache](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/aws?utm_source=gopls#NewCredentialsCache) then the SASL option looks like so:

```go
clientName := "my-app"

// read AWS from local profile:
awsConf, err := config.LoadDefaultConfig(ctx)
if err != nil {
	panic(err)
}

// AWS cached credentials like so:
credProvider := aws.NewCredentialsCache(
				awsConf.Credentials,
				func(options *aws.CredentialsCacheOptions) {
					options.ExpiryWindow = 20 * time.Second
					options.ExpiryWindowJitterFrac = 0.5
				},
			)

// the SASL option looks like this:
kgo.SASL(
	aws.ManagedStreamingIAM(func(ctx context.Context) (aws.Auth, error) {
		val, err2 := credProvider.Retrieve(ctx)
		if err2 != nil {
			return aws.Auth{}, err2
		}
		return aws.Auth{
			AccessKey:    val.AccessKeyID,
			SecretKey:    val.SecretAccessKey,
			SessionToken: val.SessionToken,
			UserAgent:    clientName,
		}, nil
	}),
)
```

my code works but randomly after 10+ hours my service caught into an infinite loop with the error looks like this:

![image](https://user-images.githubusercontent.com/816616/158736007-933d0f04-1c97-4c8b-b9c8-78116701ffb3.png)

and then when I look into the Brokers log in the AWS MSK around the same time the error first happened, this error pops (details redacted):

```
[2022-03-17T03:05:13.000+07:00] INFO [SocketServer brokerId=2] Failed re-authentication with ip-10-X-XXX-XXX.ap-southeast-1.compute.internal/INTERNAL_IP (Cannot change principals during re-authentication from IAM.arn:aws:sts::XXXX0842XXXX:assumed-role/my-app-role/1647457506320093398: IAM.arn:aws:sts::XXXX0842XXXX:assumed-role/my-app-role/1647461106753581893) (org.apache.kafka.common.network.Selector)
```

Is this caused by an invalid AWS IAM role when fetching offset/heartbeating?
Should I implement a `OnGroupManageError` hook to handle this "non-retriable" error? I can only think to restart my app in the `OnGroupManageError` hook to make it simple... but is there any way to recover from this error without needing to restart app/call panic when it happened?

Thank you for your time with this great library.

**Comment by twmb** [comment]
I'll try to help, but I haven't seen this before and I think a lot of this will be having you look into things. If you're ok with that:

That broker error looks directly from the Kafka source code, where it is ensuring the principal is unchanged. AFAICT, this is ensuring that the principal is unchanged _on the same connection_. The franz-go source supports re-authenticating.

I don't know how exactly the AccessKey vs SessionToken map to the principal, but my guess is that during this reauthentication, one of these is changing and is mapping to a different principal behind the scenes. What's weird to me is that this error should be received during sasl reauthentication, so I'd actually expect connections to fail to be opened entirely...

Do you know which of he AccessKey or SessionToken is mapping to the principal? And, if possible, can the principal be made static?

**Comment by twmb** [comment]
Any luck?

**Comment by twmb** [comment]
Rather than killing itself, would `os.Exit(..)` work?

I'm open to leaving this issue open or closing it, but if you do happen to to look into it and figure out what's up, this would be great to document.

**Comment by twmb** [comment]
Actually, going to close for now, but if you can think of a good place to put this documented workaround, lmk and we can drop it in a document somewhere (otherwise this issue will be good for history when people search). And, if you figure out what is up, that'd be awesome. Thank you!

**Comment by twmb** [comment]
Related issue: #205

**Comment by twmb** [comment]
This may be related to #205, and may be fixed now with https://github.com/twmb/franz-go/pull/242/commits/eb6e3b584d88d0bb32c8d4a856f456e2b2eae913 which is released in v1.9.1.

**Comment by twmb** [comment]
I'll start injecting group errors into polling just once in the next release with this PR: https://github.com/twmb/franz-go/pull/387
